### ISSUE TITLE
Allow ?? as a Notheing-coalescing operator

### ISSUE THREAD

**Issue by MohammadHamdyGhanem** [dotnet/vblang#381]
While writing this:
```
Function FindTSI(frm As Form) As ToolStripItem
   Return If(If(CheckToolStrip(frm.ContextMenuStrip),
                CheckToolStrip(frm.MainMenuStrip)),
                   CheckControls(frm))
End Function
```

I wondered why doesn't VB.NET use the `??` operator instead? It is more readable and used in C#. It is not strange, since VB used the `?` like in `x?.y`
The above code will became:
```
Function FindTSI(frm As Form) As ToolStripItem
   Return CheckToolStrip(frm.ContextMenuStrip) ??
             CheckToolStrip(frm.MainMenuStrip) ??
                 CheckControls(frm)
End Function
```

Which is shorter and clearer!

**Comment by jrmoreno1** [comment]
it would be more VBish IMO to overload OrElse as a null coalescing operator.

    Dim nullStr as String
    Dim str = nullStr OrElse "Oh, Happy Day"

?? is too c#ish.

**Comment by rskar-git** [comment]
Note: VB already has the coalesce operator `If(x,y)` - the If Operator Called with Two Arguments. If `x` IsNot Nothing, it returns `x`, else it returns 'y'. See https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/operators/if-operator.

**Comment by MohammadHamdyGhanem** [comment]
@rskar-git  
Please don't respond before reading the hole topic. I am suggesting the ?? instead of If() as shown in the code samples.
Thanks

**Comment by MohammadHamdyGhanem** [comment]
@jrmoreno1
OrElse can't have this new job, because it will lead to ambiguity. One example:
```
        Dim x As Byte? = Nothing
        Dim y As Byte? = 2
        Dim Z = x OrElse y
```

Now, this will make Z a boolean with the value true. But with your idea, z will be byte with the value 2. It can break some existing code!

**Comment by MohammadHamdyGhanem** [comment]
Maybe:
```
Function FindTSI(frm As Form) As ToolStripItem
  Return CheckToolStrip(frm.ContextMenuStrip) coalesce
   CheckToolStrip(frm.MainMenuStrip) coalesce
   CheckControls(frm)
End Function
```

**Comment by MohammadHamdyGhanem** [comment]
But again,  vb programmers are familiar with the `?` operator which checks if an object is nothing, so they will be familiar to `??` too.

**Comment by MohammadHamdyGhanem** [comment]
@rskar-git 
try to cascade 5 objects with If() !
` If(a, If(b, If(c, If(d, e))))`

Isn't it better to write it as:
`a ?? b ?? c ?? d ?? e`
?
Put a real expression instead of a, b, c, d, and e and compare.

**Comment by jrmoreno1** [comment]
@MohammadHamdyGhanem : your example compiles with option Strict Off, I would not consider that a barrier.  Let it do whatever it would do with Strict Off.

**Comment by MohammadHamdyGhanem** [comment]
I and other millions use option strict off, which is one of vb beauty and power. You should design on the worst case.

**Comment by jrmoreno1** [comment]
@MohammadHamdyGhanem: the code would compile to different IL depending on the Option statement, I don’t see that as a problem.  There are a lot of places where that is the case, does `Function() "this"="THIS"` return true or false?  It returns both, depending upon projects/files Option Text setting.

Strict Off is not a source of either beauty or power, it occasionally is a useful trade off between brevity and explicitness. About 80% of its usefulness could be achieved by making everything assignable to String.  It never provides a way to do anything that can’t be done otherwise (although you occasionally have to go beyond the core language to do so, mainly in relation to comparison operators).

**Comment by MohammadHamdyGhanem** [comment]
One possible workarround if:
```
    Function Coalesce(Of T)(ParamArray objects() As T) As T
        For Each o In objects
            If o IsNot Nothing Then Return o
        Next
    End Function
```

```
Function FindTSI(frm As Form) As ToolStripItem
   Return Coalesce(CheckToolStrip(frm.ContextMenuStrip),
                CheckToolStrip(frm.MainMenuStrip),
                   CheckControls(frm))
End Function
```

The problem is that one has to add the Coalesce function to each project!

**Comment by Nukepayload2** [comment]
@MohammadHamdyGhanem 
Your `Function Coalesce` is slightly different from the If(,) expression.
All expressions in the parameter list of `Coalesce` are evaluated before the call to `Coalesce`.

__VB__
```vbnet
Module Program

    Sub Main()
        Console.WriteLine("#Region ""Function Coalesce""")
        Dim result = Coalesce(Expr1, Expr2, Expr3)
        Console.WriteLine(result)
        Console.WriteLine("#End Region
")

        Console.WriteLine("#Region ""If(,) expression""")
        result = If(Expr1(), If(Expr2(), Expr3()))
        Console.WriteLine(result)
        Console.WriteLine("#End Region")
    End Sub

    Function Expr1() As String
        Console.WriteLine("Expression 1 executed")
        Return "Expr1 Has value"
    End Function

    Function Expr2() As String
        Console.WriteLine("Expression 2 executed")
        Return Nothing
    End Function

    Function Expr3() As String
        Console.WriteLine("Expression 3 executed")
        Return Nothing
    End Function

    Function Coalesce(Of T As Class)(ParamArray objects() As T) As T
        For Each o In objects
            If o IsNot Nothing Then Return o
        Next
        Return Nothing
    End Function

    Function Coalesce(Of T As Structure)(ParamArray objects() As T?) As T?
        For Each o In objects
            If o IsNot Nothing Then Return o
        Next
    End Function
End Module
```

The program generates the following output:
```
#Region "Function Coalesce"
Expression 1 executed
Expression 2 executed
Expression 3 executed
Expr1 Has value
#End Region

#Region "If(,) expression"
Expression 1 executed
Expr1 Has value
#End Region
```

**Comment by MohammadHamdyGhanem** [comment]
@Nukepayload2 
That is true. So, it will be better if vb uses it as an expression not a function. The problem is that if(x, y, z) has a special job, so we can't ask for more of if. So the ideas till now are:
1- use?? , oe Coalesce  as operators.

OrElse can be confusing. It may mean return the second if the first is false notjust nothing. So it is better reserved for booleans.
2- use Coalesce as an expression allowing any no of params.

**Comment by MohammadHamdyGhanem** [comment]
#382

**Comment by pricerc** [comment]
What you're really after is a short-circuiting Coalesce operator that's the same as the null-coalescing 'If' operator, but allows for multiple potentially null arguments.

I have a lot of sympathy for that.

But it should be: `dim x = Coalesce(a, b, c, d, e)` not `dim x = a ?? b ?? c ?? d ?? e`, since this is VB not /insert trendy language here/

(I was sure there was a proposal for this, but I couldn't find it).

**Comment by MohammadHamdyGhanem** [comment]
@pricerc 
Ok. This  is a discussion  to reach an agreement olabout the best syntax,  so I agree with you. Hope that vb team accept the concept.

**Comment by JustNrik** [comment]
I prefer `Or` instead of `OrElse`, `OrElse` doesn't make any sense since it's used for short-circuit booleans, which has nothing to do with null coalesing.

```vb
Public Sub Foo(Optional obj As SomeType = Nothing)
    Dim someObj = obj Or New SomeType()
    ' Do stuff with someObj
End Sub
```
`??` is too C#-ish as stated above, `OrElse` makes no sense so I suggest `Or` instead.

**Comment by zspitz** [comment]
@JustNrik

> `OrElse` doesn't make any sense since it's used for short-circuit booleans, which has nothing to do with null coalesing

`OrElse` would express short-circuiting null coalescing.

**Comment by JustNrik** [comment]
> 
> 
> @JustNrik
> 
> > `OrElse` doesn't make any sense since it's used for short-circuit booleans, which has nothing to do with null coalesing
> 
> `OrElse` would express short-circuiting null coalescing.

It makes no sense. short-circuiting is a boolean operation semantic, it has nothing to do with null coalesing.

**Comment by zspitz** [comment]
@JustNrik Isn't a short-circuiting null coalesce [what we're talking about here](https://github.com/dotnet/vblang/issues/381#issuecomment-466760608)? For the following code, using a fictional `Coalesce` operator:
```vb
Dim result = Coalesce(
    MayReturnNull1(),
    MayReturnNull2(),
    MayReturnNull3()
)
```
I don't want successive function calls to be evaluated if `MayReturnNull` returns null.

**Comment by JustNrik** [comment]
> 
> 
> @JustNrik Isn't a short-circuiting null coalesce [what we're talking about here](https://github.com/dotnet/vblang/issues/381#issuecomment-466760608)? For the following code, using a fictional `Coalesce` operator:
> 
> ```vbnet
> Dim result = Coalesce(
>     MayReturnNull1(),
>     MayReturnNull2(),
>     MayReturnNull3()
> )
> ```
> 
> I don't want successive function calls to be evaluated if `MayReturnNull` returns null.

Oh okay, you're right. Isn't it too verbose, though?
```vb
Dim stuff = a OrElse b OrElse c OrElse d OrElse e OrElse f ' and so...
``` I could be simplified to `Or` and still be short-circuit as it stops in the first not null anyways

**Comment by zspitz** [comment]
@JustNrik 

> It could be simplified to `Or` and still be short-circuit as it stops in the first not null anyways

It could be, but that would mean different short-circuit behavior between the boolean `Or` and the proposed null-coalescing `Or`. A little verbosity here is good.

**Comment by franzalex** [comment]
While we're on this subject, let me add my two cents.

First off, having to get the first non null value from now than two values using ```If(,) ``` is a major pain and easily becomes a mess both to read and to write too due to the reasons described above. For that reason, I'm open to the addition of a new function or operator which simplifies the process.

I lean more towards the addition of an operator since it's more easily read and understood. I also go ahead to suggest ```OrIfNull``` be the new keyword for that operator.

    Dim value = MayBeNull1() OrIfNull MayBeNull2() OrIfNull MayBeNull3()

Failing that, I think a function with a different name will serve the purpose as well. We will probably have come up with a different name than the suggested ```Coalesce``` in my opinion. Something more descriptive of the purpose such as **```FirstWithValue```**.

**Comment by MohammadHamdyGhanem** [comment]
@franzalex 
Ok, we are agree with the concept,  so it any good naming suggestion is ok.
Maybe we can use Select as the multi param if. It is short, familiar.  reserved and analogous to if:

```
 Dim x = If (a, b)
 Dim y = Select(a, b, c, d, e)
```

**Comment by rskar-git** [comment]
@franzalex

FWIW, I prefer `Coalesce`. It's already a fixture of SQL, and it's been a bona fide English word for centuries.

It would be a rare VB programmer who won't also get exposure to SQL early on.

@MohammadHamdyGhanem, my apologies for my too-hasty response. As far as `If(x,y)` goes, I guess it's never bothered me much since it's so very rare I would every need to sort through more than several values. Mostly it's been a matter of two or three values for my own needs; but I can see if one has a big list of possibilities to consider, then things can get cumbersome indeed.

I'm not a fan of `Select` for this purpose; because of the `Select Case...` statement, it would seem to me that any reuse of `Select` should be one where the first parameter is a selector - such as maybe for #160.

**Comment by zspitz** [comment]
To sum up, there are two possible advantages under discussion to a new operator:

1. Short-circuiting behavior -- the arguments will only be evaluated until a non-null argument is found
2. Easier nesting -- `x OrElse y OrElse z` instead of `If(If(x, y), z)`

over a function (1), or multiple levels of the existing `If` (2).

I think using `-Else` to express short-circuiting fits in with the other places where short-circuiting is used, and `OrElse` in general looks fine to me.

(NB. I thought that only VB.NET has non-short-circuiting boolean operators, but it turns out C# has them too; you can use the bitwise operators in a boolean context as well:
```csharp
bool result = TrueOrFalse1() ? TrueOrFalse2() ? TrueOrFalse3();
```
will evaluate all three function calls.)

**Comment by craigajohnson** [comment]
Great concept, please implement with whatever reasonable syntax makes sense!

For the record, what is the problem with expanding If(a, b, c, …)? We already use If and know its intent. Don't have a problem with Coalesce either but If is already there...

**Comment by MohammadHamdyGhanem** [comment]
@rskar-git
No need for apologies. 
I was writing c# code where I needed to test if the mouse is currently over any ToolStripItem. I had a problem detecting mouse click over an external video window, especially when a main menu or a context menu appears over the video window! I wrote code to detect which ToolStripItem is under the mouse when it is clicked, which led me to the situation where I should return the non-null ToolStripItem from different calls to search function. 
In C#, using ?? operator makes it easy to test as many expressions as you want:
```C#
ToolStripItem CheckControls(Control parent)
{            
     foreach (Control c in parent.Controls)
     {
         if (c.Visible)
         {
            var t = CheckToolStrip(c as ToolStrip) ?? 
                CheckToolStrip(c.ContextMenuStrip) ??
                CheckControls(c);

            if (t != null)
                return t;
         }
     }

     return null;
}
```
 When I wrote a VB.NET version of the the search Function, The code was longer and less clearer 
```VB.NET
Function CheckControls(parent As Control) As ToolStripItem
   For Each c As Control In parent.Controls
       If c.Visible Then
          Dim t = If(CheckToolStrip(TryCast(c, ToolStrip)),
                     If(CheckToolStrip(c.ContextMenuStrip),
                        CheckControls(c)))
          If t IsNot Nothing Then Return t
       End If
   Next c

   Return Nothing
End Function
```
so I wrote this proposal. It is logical that a portion of millions of VB.NET programmers can face scenarios like that.

**Comment by rskar-git** [comment]
Yeah, it would seem coding-style is the only way to make things livable right now.

    Function CheckControls(parent As Control) As ToolStripItem
        For Each c As Control In parent.Controls
        If c.Visible Then
            Dim t = If(
                CheckToolStrip(TryCast(c, ToolStrip)), If(
                CheckToolStrip(c.ContextMenuStrip), 
                CheckControls(c)
                ))
            If t IsNot Nothing Then Return t
        End If 
        Next c

        Return Nothing
    End Function

    ' If we had Coalesce, then "Dim y = Coalesce(a, b, c, d, e)", but right now it's:
    ' Dim y = If(a, If(b, If(c, If(d, e)))). By way of coding-style it could be:
    Dim y = If(
        a, If(
        b, If(
        c, If(
        d,
        e
        ))))

If we had to add another possibility `f`:

    Dim y = If(
        a, If(
        b, If(
        c, If(
        d, If(  ' Add If(
        e,      ' add comma
        f       ' The new possibility
        )))))   ' One more close-parenthesis

**Comment by rskar-git** [comment]
The only other option is to forsake the code-golf in favor of clarity:

    Function CheckControls(parent As Control) As ToolStripItem
        For Each c As Control In parent.Controls
            If c.Visible Then
            
                Dim t = CheckToolStrip(TryCast(c, ToolStrip))
                If t IsNot Nothing Then Return t
                
                t = CheckToolStrip(c.ContextMenuStrip)
                If t IsNot Nothing Then Return t
                
                t = CheckControls(c)
                If t IsNot Nothing Then Return t
                
            End If 
        Next c

        Return Nothing
    End Function

Or like the above but with a lambda:

    Function CheckControls(parent As Control) As ToolStripItem
        For Each c As Control In parent.Controls
            If c.Visible Then
                Dim t = (Function()
                    Dim o = CheckToolStrip(TryCast(c, ToolStrip))
                    If o Is Nothing Then o = CheckToolStrip(c.ContextMenuStrip)
                    If o Is Nothing Then o = CheckControls(c)
                    Return o
                End Function)()
                If t IsNot Nothing Then Return t                
            End If 
        Next c
        Return Nothing
    End Function

Or maybe this is optimal:

    Function CheckControls(parent As Control) As ToolStripItem
        For Each c As Control In parent.Controls
            If c.Visible Then
                Dim t = (Function()
                             Dim o = CheckToolStrip(TryCast(c, ToolStrip))
                             If o Is Nothing Then
                                 o = CheckToolStrip(c.ContextMenuStrip)
                                 If o Is Nothing Then
                                     o = CheckControls(c)
                                 End If
                             End If
                             Return o
                         End Function)()
                If t IsNot Nothing Then Return t
            End If
        Next c
        Return Nothing
    End Function

**Comment by zspitz** [comment]
@rskar-git Isn't there a happy medium:
```
Function CheckControls(parent As Control) As ToolStripItem
    For Each c As Control In parent.Controls
        If c.Visible Then
            Dim t = CheckToolStrip(TryCast(c, ToolStrip))
            t = If(t, CheckToolStrip(c.ContextMenuStrip))
            t = If(t, CheckControls(c))
            If t Is Nothing Then Contiue
            Return t
        End If 
    Next c

    Return Nothing
End Function

**Comment by rskar-git** [comment]
@zspitz

I would suppose the `t = If(t, ...)` pattern would work out fine too. :)

The only question which remains is one of performance - would the optimizations of the compiler and/or JIT take notice of the moments of extraneous assignments? (Generally speaking, write operations are slower than read ops.) I would have to do some speed tests to know for sure.

**Comment by MohammadHamdyGhanem** [comment]
If #384  is implemented, we can write:
```
The only other option is to forsake the code-golf in favor of clarity:
Function CheckControls(parent As Control) As ToolStripItem
    For Each c As Control In parent.Controls
        If c.Visible Then        
            If CheckToolStrip(TryCast(c, ToolStrip)) IsNot Nothing Then Return _CheckToolStrip
            If CheckToolStrip(c.ContextMenuStrip) IsNot Nothing Then Return _CheckToolStrip
            If CheckControls(c) IsNot Nothing Then Return _CheckControls            
        End If 
    Next c

    Return Nothing
End Function
```

**Comment by VBAndCs** [comment]
The Coalesce function that @ghost proposed can be modified to like this:
```VB.NET
    Function Coalesce(Of T)(ParamArray expressions() As Func(Of T)) As T
        For Each exp In expressions
            Try
                Dim v = exp()
                If v IsNot Nothing Then Return v
            Catch
            End Try
        Next
        Return Nothing
    End Function
```

this will prevent expression from being evaluated unless needed, put they have to be wraped in lambda expressions:

```VB.NET
Function FindTSI(frm As Form) As ToolStripItem
   Return Coalesce(
                Function() CheckToolStrip(frm.ContextMenuStrip),
                Function() CheckToolStrip(frm.MainMenuStrip),
                Function() CheckControls(frm))
End Function
```

Which makes using nested If() seem better :)
